src/zap.ts

```diff
@@ -0,0 +1,55 @@
+import { ChainId, isEthereum, isFantom } from "./chain";
+import { EthAddress, ZeroAddress } from "./helpers";
+import { Token, VaultMetadataOverrides } from "./types";
+
+export type ZapInWith = keyof Token["supported"];
+
+type ZapInDetails = { isZapInSupported: boolean; zapInWith?: ZapInWith };
+
+type ZapInArgs = {
+  chainId: ChainId;
+  token?: Partial<Token>;
+  vaultMetadata: VaultMetadataOverrides | null;
+};
+
+export function getZapInDetails({ chainId, token, vaultMetadata }: ZapInArgs): ZapInDetails {
+  if (!token?.supported || !vaultMetadata?.zapInWith) {
+    return { isZapInSupported: false };
+  }
+
+  if (isEthereum(chainId)) {
+    if (token.address === EthAddress) {
+      return { isZapInSupported: Boolean(token.supported.zapper), zapInWith: "zapperZapIn" };
+    }
+
+    const { zapInWith } = vaultMetadata;
+
+    if (zapInWith && isValidZap(zapInWith, token.supported)) {
+      const isZapInSupported = Boolean(token.supported.zapper) && Boolean(token.supported[zapInWith]);
+      return { isZapInSupported, zapInWith };
+    }
```

Comment:
Could we remove the `IF` if we manage zapInWith accordingly ?
```suggestion
      const isZapInSupported = Boolean(token.supported.zapper) && Boolean(token.supported[zapInWith]);
      return { isZapInSupported, zapInWith: isZapInSupported ? zapInWith : null };
```

Reply:
I think it doesn't hurt that we check that a valid `zapInWith` is being given, what would be the benefit of removing it?

Reply:
No benefits other than less code. Ill leave it up to you :)

Reply:
I'm more comfortable leaving the check in just in case as it's somewhat of a dynamic value and TS cannot guarantee that it's of the correct type as it just gets a `string` right now.

Reply:
Forgot to mention here that `isValidZap(zapInWith, token.supported)` is also a type guard and `token.supported[zapInWith]` will have a type error without it

<img width="1857" alt="Screen Shot 2022-04-24 at 12 48 15 pm" src="https://user-images.githubusercontent.com/78794805/164953953-5dff6463-619e-4642-a399-cf72ceb3588d.png">